neqo-transport/src/quic_datagrams.rs

```diff
@@ -10,37 +10,61 @@ use crate::frame::{FRAME_TYPE_DATAGRAM, FRAME_TYPE_DATAGRAM_WITH_LEN};
 use crate::packet::PacketBuilder;
 use crate::recovery::RecoveryToken;
 use crate::stats::FrameStats;
-use crate::{ConnectionEvents, Error, Res};
+use crate::{events::OutgoingDatagramOutcome, ConnectionEvents, Error, Res};
 use neqo_common::Encoder;
 use std::cmp::min;
+use std::collections::VecDeque;
 use std::convert::TryFrom;
+use std::ops::Deref;
 
 pub const MAX_QUIC_DATAGRAM: u64 = 65535;
 
+struct QuicDatagram {
+    data: Vec<u8>,
+    id: Option<u64>,
+}
+
+impl QuicDatagram {
+    fn id(&self) -> Option<u64> {
+        self.id
+    }
+}
+
+impl Deref for QuicDatagram {
+    type Target = [u8];
+    #[must_use]
+    fn deref(&self) -> &[u8] {
+        &self.data[..]
+    }
+}
+
 pub struct QuicDatagrams {
     /// The max size of a datagram that would be acceptable.
     local_datagram_size: u64,
     /// The max size of a datagram that would be acceptable by the peer.
     remote_datagram_size: u64,
+    max_queued_outgoing_datagrams: usize,
     /// The max number of datagrams that will be queued in connection events.
     /// If the number is exceeded, the oldest datagram will be dropped.
-    max_queued_datagrams: usize,
+    max_queued_incoming_datagrams: usize,
     /// Datagram queued for sending.
-    datagram: Option<Vec<u8>>,
+    datagrams: VecDeque<QuicDatagram>,
     conn_events: ConnectionEvents,
 }
 
 impl QuicDatagrams {
     pub fn new(
         local_datagram_size: u64,
-        max_queued_datagrams: usize,
+        max_queued_outgoing_datagrams: usize,
+        max_queued_incoming_datagrams: usize,
         conn_events: ConnectionEvents,
     ) -> Self {
         Self {
             local_datagram_size,
             remote_datagram_size: 0,
-            max_queued_datagrams,
-            datagram: None,
+            max_queued_outgoing_datagrams,
+            max_queued_incoming_datagrams,
+            datagrams: VecDeque::new(),
```

Comment:
```suggestion
            datagrams: VecDeque::with_capacity(max_queued_outgoing_datagrams),
```

The extra overhead of this allocation will be tiny and we can avoid the overhead of allocation on the first call.